sprint_9-solution-http-api

max7358 opened the pull request.

max7358 committed:
http api imp
rest tests coverage

Sla-als commented on src/com/yandex/app/http/HttpTaskServer.java:
```diff
@@ -0,0 +1,41 @@
+        }
+    }
+
+    public void stop() {
```
Стоит добавить проверку на null для переменной server в методе stop(), чтобы избежать возможных NullPointerException при попытке остановить сервер, который не был инициализирован.

max7358 replied:
добавил проверку на null

Sla-als commented on src/com/yandex/app/http/HttpTaskServer.java:
```diff
@@ -0,0 +1,41 @@
+            server.createContext("/prioritized", new PrioritizedHandler(taskManager));
+            server.start();
+        } catch (IOException e) {
+            throw new RuntimeException(e);
```
Все catch блоки лучше расширить, добавив логирование ошибок. Это позволит лучше понять контекст возникновения ошибок

max7358 replied:
поменял на кастомный эксепшен

Sla-als commented on src/com/yandex/app/http/handler/BaseHttpHandler.java:
```diff
@@ -0,0 +1,97 @@
+        h.close();
+    }
+
+    protected void sendNotFound(HttpExchange h, String text) throws IOException {
```
Методы sendNotFound и другие методы отправки ответов дублируют код. Можно рассмотреть возможность создания утилитарного метода, который бы принимал код ответа и сообщение, чтобы уменьшить дублирование.

Sla-als replied:
В случае использования нескольких типов ошибок, стоит создать кастомные исключения, которые более точно описывают проблему, вместо использования общего NotFoundException.

max7358 replied:
оптимизировал код

max7358 replied:
разные ошибки кидают разные кастомные исключения, в сообщение написано что не так:
```
        try {
            Task task = gson.fromJson(requestBody, Task.class);
            if (task.getId() > 0) {
                taskManager.updateTask(task);
            } else {
                taskManager.addTask(task);
            }
            sendCreated(exchange);
        } catch (ValidationException e) {
            sendHasInteractions(exchange, e.getMessage());
        } catch (NotFoundException e) {
            sendNotFound(exchange, e.getMessage());
        }
```

max7358 replied:
текст ошибки для ValidationException  как пример:
Error: task id:8 - 2024-04-02T14:21:10.000000020Z : 2024-04-02T14:52:10.000000020Z
intercepts with existing task id:8 - 2024-04-02T14:21:10.000000020Z : 2024-04-02T14:52:10.000000020Z

max7358 replied:
текст ошибки для ValidationException как пример:
Error: task id:9 - 2024-04-02T14:22:10.000000020Z : 2024-04-02T14:52:10.000000020Z
intercepts with existing task id:8 - 2024-04-02T14:21:10.000000020Z : 2024-04-02T14:52:10.000000020Z

Sla-als replied:
Отлично! Спасибо за комменатрии

Sla-als commented on test/service/FileBackedTaskManagerTest.java:
Тесты хорошо структурированы, но стоит добавить больше тестов для негативных сценариев и проверки поведения системы при невалидных входных данных.

max7358 replied:
Это касательно всех тестов или конкретно "файловых" ?
добавил тест, если файла не существует.
Касательно валидности входных данных, наставник писал что на данном этапе считаем что получаем только валидные данные:
_Как правило в приложениях есть валидация и хорошо, что ты о ней задумался, но это тема будущих спринтов)
После ревью можешь попытаться сделать самостоятельно, в будущем разберем как это принято делать, будет с чем сравнить)_

Sla-als commented on src/com/yandex/app/http/adapter/DurationTypeAdapter.java:
Все адаптеры должны обрабатывать null значения корректно для предотвращения возможных NullPointerException.

max7358 replied:
поправил

Sla-als reviewed (commented):
В целом, ваш код хорошо структурирован и логичен. Вы проделали большую работу по созданию чистой архитектуры HTTP-сервера с четким разделением ответственности. Я  оставил несколько замечаний:

max7358 committed:
review fixes:
method stop() null check
replaced RuntimeException(e) with custom one
response methods code improved
test for FileBackedTaskManager

max7358 committed:
test for FileBackedTaskManager - test fix

Sla-als commented:
Отлично! Видно, что проделана большая работа на спринте! Поздравляю с заслуженным зачетом)
